From 2d0f447b61b8bbc7eaf9c9f7c7c2caf7c486a645 Mon Sep 17 00:00:00 2001
From: Ranjani Vaidyanathan <ranjani.vaidyanathan@nxp.com>
Date: Tue, 21 Jan 2020 16:12:03 -0600
Subject: [PATCH 24/37] MLK-23258-1 soc:imx8qm/imx8qxp Fix PCIE suspend/resume
 issue

commit 9f9f5a3c7c413928daae24f5b3f12f2bc25be831 from
https://source.codeaurora.org/external/imx/linux-imx.git

The power domain driver should ensure all the child power domains
are powered off before powering down the parent.

Signed-off-by: Ranjani Vaidyanathan <ranjani.vaidyanathan@nxp.com>
(cherry picked from commit fa35a834810156c35a06c9ea2c962ee02063eed9)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/soc/imx/pm-domains.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/drivers/soc/imx/pm-domains.c b/drivers/soc/imx/pm-domains.c
index b9861db60239..2ffcce350c21 100644
--- a/drivers/soc/imx/pm-domains.c
+++ b/drivers/soc/imx/pm-domains.c
@@ -126,9 +126,10 @@ static int imx8_pd_power(struct generic_pm_domain *domain, bool power_on)
 		 */
 		while (!list_empty(&cur_domain->slave_links)) {
 			struct gpd_link *link;
-
 			pd = container_of(cur_domain, struct imx8_pm_domain, pd);
-			if ((cur_domain == domain) || (!cur_domain->device_count)) {
+			if ((cur_domain == domain) ||
+				((!cur_domain->device_count) &&
+				(atomic_read(&cur_domain->sd_count) <= 1))) {
 				sci_err = sc_pm_set_resource_power_mode(pm_ipc_handle,
 						pd->rsrc_id,
 						(pd_state) ? SC_PM_PW_MODE_OFF : SC_PM_PW_MODE_LP);
@@ -137,7 +138,8 @@ static int imx8_pd_power(struct generic_pm_domain *domain, bool power_on)
 							pd->rsrc_id, sci_err);
 					return -EINVAL;
 				}
-			}
+			} else
+				break;
 
 			list_for_each_entry(link, &cur_domain->slave_links, slave_node)
 				master = link->master;
@@ -147,7 +149,9 @@ static int imx8_pd_power(struct generic_pm_domain *domain, bool power_on)
 		/* Fix the state for the top parent or a node that has no slave domains */
 		pd = container_of(cur_domain, struct imx8_pm_domain, pd);
 		if ((cur_domain == domain) ||
-			((pd->rsrc_id != SC_R_NONE) && (!cur_domain->device_count))) {
+			((pd->rsrc_id != SC_R_NONE) &&
+			(!cur_domain->device_count) &&
+			(atomic_read(&cur_domain->sd_count) <= 1))) {
 			sci_err = sc_pm_set_resource_power_mode(pm_ipc_handle,
 						pd->rsrc_id,
 						(pd_state) ? SC_PM_PW_MODE_OFF : SC_PM_PW_MODE_LP);
-- 
2.17.1

